libraries/timeslider/src/components/TimeSlider/TimeSlider.tsx

```diff
@@ -0,0 +1,101 @@
+import React, { ComponentType, ReactElement, useCallback } from "react";
+import { LayoutElementProperties } from "@vertigis/web/components";
+import EsriTimeSlider from "@arcgis/core/widgets/TimeSlider";
+import TimeSliderModel from "./TimeSliderModel";
+import {
+    createEsriMapWidget,
+    MapWidgetProps,
+} from "@vertigis/web/ui/esriUtils";
+import "./TimeSlider.css";
+import { useWatch, useWatchAndRerender } from "@vertigis/web/ui";
+import WebMap from "@arcgis/core/WebMap";
+
+export type TimeSliderProps = MapWidgetProps<TimeSliderModel, EsriTimeSlider>;
+const TimeSliderWrapper: ComponentType<TimeSliderProps> = createEsriMapWidget<
+    TimeSliderModel,
+    EsriTimeSlider
+>(EsriTimeSlider, undefined, true);
+
+const TimeSlider = (
+    props: LayoutElementProperties<TimeSliderModel>
+): ReactElement => {
+    // A unique DOM ID to be used for a11y purposes.
+    const model: TimeSliderModel = props.model;
+
+    // Any time the Time Slider model properties change, we need to re-render
+    // the component so our UI is consistent, as well as set the Esri widget
+    // configuration.
+    useWatchAndRerender(model, [
+        "layout",
+        "loop",
+        "playRate",
+        "mode",
+        "overrideStops",
+        "timeInterval",
+        "timeIntervalUnit",
+        "timeVisible",
+        "map.map",
+    ]);
+    useWatch(
+        model,
+        [
+            "layout",
+            "loop",
+            "playRate",
+            "mode",
+            "overrideStops",
+            "timeInterval",
+            "timeIntervalUnit",
+            "timeVisible",
+        ],
+        (newValue, _oldValue, propertyName) => {
+            if (model.widget) {
+                model[propertyName] = newValue;
+            }
+        }
+    );
+    useWatch(
+        model,
+        "map.map",
+        (newMap) => {
+            // Only update the model if the web map has changed.
+            if (model.widget) {
+                model.widget.stop();
+                // Check the web map for existing time slider config
+                model.updateTimeSliderWidget(model.widget, newMap);
+            }
+        },
+        [model.map.view]
+    );
+
+    const onWidgetCreated = useCallback(
+        (widget: EsriTimeSlider) => {
+            const map = model.map;
+
+            // Synchronize values from model.
+            if (!map && !map.map) {
+                return;
+            }
+
+            const webMap = map.map as WebMap;
+            model.updateTimeSliderWidget(widget, webMap);
+        },
+        [model]
+    );
+    const onWidgetDestroyed = useCallback(() => {
+        // Remove the time extent filter on the map view.
+        model.map.view.timeExtent = null;
+        model.widget = null;
+    }, [model]);
+
+    return (
+        <TimeSliderWrapper
+            model={props.model}
+            preloadChildren={() => Promise.resolve()}
```

Comment:
The `preloadChildren` property was added because it's a non-optional property in LayoutElementProperties (which `@vertigis/web/ui/esriUtils/MapWidgetProps` extends) and I saw the same pattern used in a couple places in the web viewer repo. Would it make sense to just make this property optional in web?

Reply:
Oh, I see. It's non-optional because the layout engine guarantees that this will always be passed in. Layout elements are not meant to be created directly, but rather are created by the layout engine. Since the wrapper returned from `createEsriMapWidget()` is a layout element, then if it is wrapped by a new component (like it is here), all properties that belong to LayoutElementProperties and basic DOM properties must be forwarded to the  widget wrapper using `{...other}`. Otherwise, none of the various layout attributes like `width`, `height`, `grow`, and so on will work properly, and it will be missing `gcx-id` attribute (soon to be `data-layout-id`), important CSS classes, and lots more. This should remove the need to put this stub in place. Any non-test code in Web that is also doing this is doing something incorrect and should be fixed.

Reply:
Props fixed.